packages-exp/auth-exp/src/core/util/assert.ts

```diff
@@ -44,6 +44,92 @@ export function assert(
   }
 }
 
+type TypeExpectation = Function | string | MapType;
+
+interface MapType extends Record<string, TypeExpectation | Optional> {}
+
+class Optional {
+  constructor(readonly type: TypeExpectation) {}
+}
+
+export function opt(type: TypeExpectation): Optional {
+  return new Optional(type);
+}
+
+/**
+ * Asserts the runtime types of arguments. The 'expected' field can be one of
+ * a class, a string (representing a "typeof" call), or a record map of name
+ * to type. Furthermore, the opt() function can be used to mark a field as
+ * optional. For example:
+ *
+ * function foo(auth: Auth, profile: {displayName?: string}, update = false) {
+ *   assertTypes(arguments, [AuthImpl, {displayName: opt('string')}, opt('boolean')]);
+ * }
+ *
+ * opt() can be used for any type:
+ * function foo(auth?: Auth) {
+ *   assertTypes(arguments, [opt(AuthImpl)]);
+ * }
+ *
+ * The string types can be or'd together, and you can use "null" as well (note
+ * that typeof null === 'object'; this is an edge case). For example:
+ *
+ * function foo(profile: {displayName?: string | null}) {
+ *   assertTypes(arguments, [{displayName: opt('string|null')}]);
+ * }
+ *
+ * @param args
+ * @param expected
+ */
+export function assertTypes(
+  args: Omit<IArguments, 'callee'>,
+  ...expected: Array<TypeExpectation | Optional>
+): void {
+  if (args.length > expected.length) {
+    fail('TODO', AuthErrorCode.ARGUMENT_ERROR);
```

Comment:
can this be debugFail? We should never hit this based on user input, right?

Reply:
Nope, for example:

```ts
function foo(auth: Auth) {
  assertTypes(arguments, AuthImpl);
}
```
should fail with this call (which is perfectly valid in javascript)
```js
foo(auth, 'some-other-parameter');
```